use PhoenixProfiler on the Endpoint

mcrumm opened the pull request.

mcrumm committed:
WIP: Add profiler Endpoint

mcrumm committed:
Simplify tests

mcrumm committed:
wip: profile

mcrumm committed:
maybe mount profile

mcrumm committed:
Consolidate profiler modules

mcrumm committed:
profile thru endpoint

mcrumm committed:
Clarify endpoint config

mcrumm committed:
docs

mcrumm committed:
docs

mcrumm committed:
remove unused key

mcrumm committed:
use PhoenixProfiler

mcrumm committed:
docs

mcrumm committed:
Fix dashboard

mcrumm committed:
Apply limit to dashboard profile list

mcrumm committed:
dashboard profile time

mcrumm committed:
ignore live_reload

mcrumm committed:
endpoint tests

mcrumm committed:
Collect exception from endpoint

mcrumm committed:
Rename plug telemetry

mcrumm committed:
Sync (dis|en)abling telemetry collector

mcrumm committed:
Stabilize profiler responses

mcrumm committed:
Update collector controls on ToolbarLive

mcrumm committed:
Fix empty toolbar render

mcrumm committed:
Merge branch 'main' into mc-endpoint

* main:
  Fix package deps
  Release v0.1.0
  Recommend profiler on telemetry supervision tree
  Add a Changelog

mcrumm committed:
Add endpoint notes to changelog

mcrumm committed:
Add dashboard limit to changelog

mcrumm committed:
Drop unused helper

leandrocp commented on lib/phoenix_profiler/toolbar/toolbar_live.ex:
```diff
@@ -65,9 +65,19 @@ defmodule PhoenixProfiler.ToolbarLive do
+    %Profile{data: %{metrics: metrics}} = profile
+
+    socket =
+      case profile.data[:exception] do
```
Seems like there's a race condition here, the toolbar is mounted before telemetry has a chance to collect the event ` [:phoenix_profiler, :endpoint, :exception]` so `profile.data[:exception]` is always null at this point.

I'm assuming that http://localhost:4000/errors/assign-not-available should populate `@exits` considering https://github.com/mcrumm/phoenix_profiler/blob/ae6ca03ee8237e51844d099562baaf5888ec33c4/lib/phoenix_profiler/endpoint.ex#L55 and https://github.com/mcrumm/phoenix_profiler/blob/ae6ca03ee8237e51844d099562baaf5888ec33c4/lib/phoenix_profiler/telemetry.ex#L43

mcrumm commented:
@leandrocp You are correct– the toolbar cannot render the error for its own request because the `Phoenix.Endpoint.RenderErrors` module takes over sending the response. To account for this, the private `late_collect/3` function on the profiler endpoint is invoked _after_ the response has been sent, which allows us to collect additional information about the request to be viewed later.

I think you are also correct about the race condition tho– we are racing the late write to ETS and that's why we are getting back profile data without the exception.

But now I think I have an idea of how to get the tests passing, stand by! :)

mcrumm committed:
Validate exceptions on tokens thru the Endpoint

The endpoint integration tests are racing the late collection
behaviour in the profiler endpoint, which sometimes results in
tests failing due to receiving outdated profile information.

We are attempting to correct this by fetching the profile data
by its token thru another endpoint request. Hopefully this will
slow things down enough to give the late write a much better chance
to finish before the data is requested.

In practice this isn't an actual problem, as in most cases (like the
Dashboard) we poll for updated information.

mcrumm committed:
Endpoint integration tests wait once for expected result

mcrumm committed:
Send real result after retry

mcrumm committed:
increase wait before retrying

mcrumm committed:
Simplify waiting on profile data for endpoint test

mcrumm committed:
increase wait timeout

mcrumm committed:
Slight wait between checks

mcrumm committed:
Add [:phoenix, :error_rendered] to telemetry events

leandrocp reviewed (approved):
Hey @mcrumm just tested this PR on a real project and it looks good to me 👍🏻

mcrumm committed:
debug test output

mcrumm committed:
drop debug statements

mcrumm committed:
normalize exception tests

mcrumm commented:
@leandrocp Thanks for testing it out! Yes, things seem to work fine but the Heisenbugs are forcing me to question some earlier decisions. I just had a nice chat with @elbow-jason who gave me some ideas that I hope to experiment with soon! :)

mcrumm committed:
Merge branch 'main' into mc-endpoint

* main:
  Update version in README
  Release v0.2.0
  Update supported LiveView versions (#65)
  Update dev.exs
  Add troubleshooting for on_mount, closes #61
  Simulate hooks on demo app
  Handle multiple body tags

mcrumm committed:
assert capture_log
